pilot/pkg/networking/plugin/authz/rbac_v2.go

```diff
@@ -0,0 +1,172 @@
+					continue
+				}
+			}
+
+			if option.forTCPFilter {
+				if err := validateBindingsForTCPFilter([]*rbacproto.ServiceRoleBinding{subject}); err != nil {
+					rbacLog.Debugf("role %s skipped, found HTTP only binding for a TCP service: %v", subject.RoleRef.Name, err)
+					continue
+				}
+			}
+			// Convert binding.
+			enforcedPrincipals, _ := convertToPrincipals([]*rbacproto.ServiceRoleBinding{subject}, option.forTCPFilter)
+
+			if len(enforcedPrincipals) == 0 {
+				rbacLog.Debugf("role %s skipped for no principals found", subject.RoleRef.Name)
+				continue
+			}
+
+			if len(enforcedPrincipals) != 0 {
+				if rbac.Policies[subject.RoleRef.Name] == nil {
+					rbac.Policies[subject.RoleRef.Name] = &policyproto.Policy{
+						Permissions: []*policyproto.Permission{permission},
+						Principals:  enforcedPrincipals,
+					}
+				} else {
+					allPrincipals := rbac.Policies[subject.RoleRef.Name].Principals
+					// Check if enforcedPrincipals has existed in the allPrincipals.
+					if !isPrincipalInPrincipals(enforcedPrincipals, allPrincipals) {
+						rbac.Policies[subject.RoleRef.Name].Principals = append(allPrincipals, enforcedPrincipals...)
+					}
+				}
+			}
+		}
+	}
+	return &http_config.RBAC{Rules: rbac}
+}
+
+// isServiceMatchedWorkload checks if the calling service is matched with the selector in the AuthorizationPolicy if
+// present.
+func (service serviceMetadata) isServiceMatchedWorkload(labels map[string]string) bool {
+	workloadLabels := model.LabelsCollection{labels}
+	return workloadLabels.HasSubsetOf(service.labels)
+}
+
+func isRbacV2(namespace string, option rbacOption) bool {
+	// Check to make sure we are not dereferencing null pointers.
+	if option.authzPolicies == nil || option.authzPolicies.NamespaceToAuthorizationConfigV2 == nil {
+		return false
+	}
+	if _, present := option.authzPolicies.NamespaceToAuthorizationConfigV2[namespace]; !present {
+		return false
+	}
+
+	// Get all AuthorizationPolicy Istio config from this namespace.
+	allAuthzPolicies := option.authzPolicies.NamespaceToAuthorizationConfigV2[namespace].AuthzPolicies
+	return len(allAuthzPolicies) > 0
+}
+
+// isConflictingRbacVersion is called when the RBAC filter is supposed to use RBAC v2 (with AuthorizationPolicy).
+// It checks to see if ServiceRoleBinding still exists. If so, return True; False otherwise.
+func isConflictingRbacVersion(namespace string, option rbacOption) bool {
+	return len(option.authzPolicies.RoleToBindingsForNamespace(namespace)) > 0
+}
+
+// isPrincipalInPrincipals returns True if the |principal| has already existed in the |principals|,
+// False otherwise.
+// This is to make sure we don't have duplicated principals for the same ServiceRole.
+// This can happen when we have a service foo that has more than one exact Subjects (in one or many
+// AuthorizationPolicy) that bind to the same ServiceRole.
+func isPrincipalInPrincipals(principal []*policyproto.Principal, principals []*policyproto.Principal) bool {
```

Comment:
I don't understand why do we need this, why do we need to make sure we don't have duplicated principals for the same ServiceRole?

Reply:
Say we have a ServiceRole `SR`, and a binding `B` that appears in 2 AuthorizationPolicy CRDs using the same `SR` and pointing to workload or service. When we convert, we will attempt to convert `B` twice for the same `SR`. When we see `B` the second time, we can ignore it. This doesn't  affect the final results of the RBAC that this service or workload will receive, but we don't need to store duplicated binding since an exact binding has already exists.

Reply:
First, if the two AuthorizationPolicy has different selector, we still need to convert its binding `B` twice, right? As they are applying to different workload.
Second, we should avoid such optimization at this stage if it's critical to affect the performance heavily.